### PR TITLE
Refactor state into config (url-encoded) and 'data/app state' 

### PR DESCRIPTION
**Pull request by bryophyta**
## What does this change?

**nb.** depends on #42 

in brief: refactor the state management so that 

1. the URL management hook (formerly 'useHistory', now 'useUrlConfig') knows more about the structure of config (app view/page, search query, and selected item id if it exists), and tries to confine serialisation to/from URL strings to the edge of the hook
2. all direct use of the `useUrlConfig` hook is contained now in the `useSearch` hook
3. it's more explicit that `useSearch` concerns data/app state that _isn't_ encoded in the URL, i.e. fetched data, loading state, errors, etc.
4. and this app state is now handled via a state machine with `useReducer`, with config/url changes handled as side-effects via `useEffect` hooks.

The broad intention is to make it so that all state changes within the app itself can be achieved by dispatching actions to the `useSearch` reducer, without needing to know about reading or writing to the URL.

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

**Review by andrew-nowak — APPROVED** [review]
lgtm! nice refactor! 

I've left a couple of small comments, the only other relate point I'll leave to be thought about: 
I'm hoping to bring in something like https://github.com/guardian/crosswordv2/blob/main/client/src/state.tsx#L165, ie. all the cards in the feed are proper `<a>`s, so can be middle-clicked, right-clicked, dragged etc. which does mean "knowing" the href (though I feel like that can be delegated back to the `configToUrl` function in some clever way... 
for a later think, and definitely not essential!

File: newswires/client/src/useSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	);
+}
+
+export const useSearch = () => {
+	const historyContext = useContext(SearchContext);
+	if (historyContext === null) {
+		throw new Error('useHistory must be used within a HistoryContextProvider');
+	}
+	return historyContext;
+};
```
[the remark this answers is not in the document]
Comment (bryophyta) [reply newswires/client/src/useSearch.tsx]: thanks! done here: https://github.com/guardian/newswires/pull/47/commits/7ed68bbba59b5393c5cb753a663ad0a773ebf4ec

File: newswires/client/src/urlState.tsx [hunk 2]
```diff
@@ -80,96 +68,59 @@ const paramsToQuerystring = (p: HistoryState['params']): string => {
...
-	currentState: HistoryState;
-	pushState: (state: HistoryState) => void;
-	replaceState: (state: HistoryState) => void;
-};
-const HistoryContext: Context<HistoryContextShape | null> =
-	createContext<HistoryContextShape | null>(null);
-
-export const HistoryContextProvider = ({ children }: PropsWithChildren) => {
-	const [currentState, setState] = useState<HistoryState>(readUrl());
+export const useUrlConfig = () => {
```
[the remark this answers is not in the document]
Comment (bryophyta) [reply newswires/client/src/urlState.tsx]: thanks! I held off partly to reduce the diff in this PR but I think you're right that it's worth making sure we inline it. Done now: 6ba973de9177596871e0e163eb6f33b35e4f6966

**Comment by bryophyta** [comment]
> ie. all the cards in the feed are proper <a>s, so can be middle-clicked, right-clicked, dragged etc.

definitely agree that this would be good to do, and I think we should be able to do this in a future PR 👍

**Pull request merged**